Added missing ProducesResponseType, aligned response types

bartlannoeye opened the pull request:
Fixes #10613

Also added the types in the ResponseType for .NetCore > 2.1

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.21/Controllers/PetsController.cs:
```diff
@@ -19,12 +19,14 @@ public PetsController(PetsRepository repository)
         }
 
         [HttpGet("{id}")]
+        [ProducesResponseType(typeof(Pet), 200)]
```
```suggestion
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.21/Controllers/PetsController.cs:
```diff
@@ -19,12 +19,14 @@ public PetsController(PetsRepository repository)
         }
 
         [HttpGet]
+        [ProducesResponseType(typeof(IEnumerable<Pet>), 200)]
```
A 200 status code and the return type are implied because `ActionResult<T>` is returned.

```suggestion
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.21/Controllers/ProductsController.cs:
```diff
@@ -21,7 +21,7 @@ public ProductsController(ProductsRepository repository)
         #region snippet_GetById
         [HttpGet("{id}")]
-        [ProducesResponseType(200)]
+        [ProducesResponseType(typeof(Product), 200)]
```
```suggestion
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.21/Controllers/ProductsController.cs:
```diff
@@ -38,6 +38,7 @@ public async Task<ActionResult<Product>> GetByIdAsync(int id)
 
         #region snippet_BindingSourceAttributes
         [HttpGet]
+        [ProducesResponseType(typeof(List<Product>), 200)]
```
```suggestion
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.22/Controllers/ProductsController.cs:
```diff
@@ -17,7 +17,7 @@ public ProductsController(ProductsRepository repository)
 
         [HttpGet("{id}")]
-        [ProducesResponseType(200)]
+        [ProducesResponseType(typeof(Product), 200)]
```
```suggestion
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.22/Controllers/ProductsController.cs:
```diff
@@ -34,6 +34,7 @@ public async Task<ActionResult<Product>> GetByIdAsync(int id)
         }
 
         [HttpGet]
+        [ProducesResponseType(typeof(List<Product>), 200)]
```
```suggestion
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.22/Controllers/ProductsController.cs:
```diff
@@ -52,7 +53,7 @@ public async Task<ActionResult<Product>> GetByIdAsync(int id)
 
         [HttpPost]
-        [ProducesResponseType(201)]
+        [ProducesResponseType(typeof(Product), 201)]
```
The type is inferred from the `T` in `ActionResult<T>`.

```suggestion
        [ProducesResponseType(201)]
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.21/Controllers/PetsController.cs:
```diff
@@ -39,6 +41,7 @@ public async Task<ActionResult<Pet>> GetByIdAsync(int id)
         }
 
         [HttpPost]
+        [ProducesResponseType(typeof(Pet), 201)]
```
```suggestion
        [ProducesResponseType(201)]
```

scottaddie commented on aspnetcore/web-api/define-controller/samples/WebApiSample.Api.21/Controllers/ProductsController.cs:
```diff
@@ -57,7 +58,7 @@ public async Task<ActionResult<Product>> GetByIdAsync(int id)
 
         [HttpPost]
-        [ProducesResponseType(201)]
+        [ProducesResponseType(typeof(Product), 201)]
```
```suggestion
        [ProducesResponseType(201)]
```

bartlannoeye commented:
@scottaddie I overlooked the paragraph that the 200 return code is implied by the ActionResult. Personally I would still add it for clarity (to other devs) showing that it returns both 200 and 404, not only 404.

However for the sake of testing, I removed ALL ProducesResponseType attributes in the PetsController for .NET Core 2.1 (not 2.2), cleaned and rebuilt the solution. All response types and codes are correctly shown in Swagger. Makes me wonder if they're all added automatically because of the code paths? If so (which would be nice), which are the cases where they still have to be added manually?

scottaddie commented:
@bartlannoeye One of the benefits of using `ActionResult<T>` is reducing "attribute soup" on actions. There's absolutely nothing wrong with being explicit. Continue to do that in your code if that's your preference. We choose not to use the attribute in the docs so that you're getting the full benefit from `ActionResult<T>`.

The 200 status code is an assumption the API is making. In other words, the API isn't intelligently scanning the code paths in the action to determine that.

bartlannoeye commented:
Thanks for the response. How am I getting the other response codes even though I removed all attributes in the .NET Core 2.1 project? If I remember correctly, configuration in the Startup to handle everything was only added in .NET Core 2.2? Am I overlooking something?

scottaddie commented:
@bartlannoeye Do you have a sample project on GitHub that reproduces what you're seeing?

Rick-Anderson commented:
@scottaddie 	is this ready to merge?

bartlannoeye commented:
Can't reproduce anymore after I restarted VS, assumes 200's now as you explained.

Rick-Anderson commented:
@scottaddie is this ready to merge?

scottaddie commented:
@bartlannoeye Thank you for all your work on this PR! Your changes will go live within the next 24 hours.
